cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/PullAppsOperation.java

```diff
@@ -16,25 +16,113 @@
 
 package io.cdap.cdap.internal.app.sourcecontrol;
 
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.assistedinject.Assisted;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.app.RunIds;
 import io.cdap.cdap.internal.operation.LongRunningOperation;
 import io.cdap.cdap.internal.operation.LongRunningOperationContext;
+import io.cdap.cdap.proto.id.ApplicationId;
 import io.cdap.cdap.proto.operation.OperationError;
+import io.cdap.cdap.proto.operation.OperationResource;
+import io.cdap.cdap.proto.operation.OperationResourceScopedError;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.sourcecontrol.SourceControlException;
+import io.cdap.cdap.sourcecontrol.operationrunner.InMemorySourceControlOperationRunner;
+import io.cdap.cdap.sourcecontrol.operationrunner.MultiPullAppOperationRequest;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
 
 /**
  * Defines operation for doing SCM Pull for connected repositories.
- * TODO(samik) implement the pull-op
  **/
 public class PullAppsOperation implements LongRunningOperation {
 
   private final PullAppsRequest request;
 
-  public PullAppsOperation(PullAppsRequest request) {
+  private final InMemorySourceControlOperationRunner scmOpRunner;
+  private final SourceControlHelper helper;
+
+  private final Set<ApplicationId> deployed;
+
+  /**
+   * Only request is passed using AssistedInject. See {@link PullAppsOperationFactory}
+   *
+   * @param request contains apps to push
+   * @param runner runs git operations. The reason we do not use
+   *     {@link io.cdap.cdap.sourcecontrol.operationrunner.SourceControlOperationRunner} rather than
+   *     concrete implementation is because the git operations should always run inMemory.
+   * @param helper provides utilities to provide app-fabric exposed functionalities.
+   */
+  public PullAppsOperation(@Assisted PullAppsRequest request,
+      InMemorySourceControlOperationRunner runner,
+      SourceControlHelper helper) {
     this.request = request;
+    this.helper = helper;
+    this.scmOpRunner = runner;
+    this.deployed = new HashSet<>();
   }
 
   @Override
   public ListenableFuture<OperationError> run(LongRunningOperationContext context) {
```

Comment:
normally the future type is whatever the return value of the corresponding blocking call would be, with any errors thrown as exceptions. It would make more sense for this to return information about all the applications that were deployed, or something along those lines.

Reply:
1. The reason I am returning `OperationError` is to support operations with partial failures i.e if the operation wants to continue processing the whole batch and accumulate any error happening for each resources. Throwing exception does not feel like a good model.
2. We don't have any usage for returning the operated/created resources here. As we want to know exactly what resources were changed in case the runtime crashed, we want to update the resources in operation metadata if and when the resource is changed through the `updateResources` method in the `LongRunningOperationContext `